core/objects.go

```diff
@@ -146,6 +147,38 @@ type ValidationRecord struct {
 	AddressesTried []net.IP `json:"addressesTried,omitempty"`
 }
 
+type RehydratedValidationRecordFields struct {
+	Hostname string
+	Port     string
+}
+
+// TransformURLIntoHostnameAndPort takes a URL string as input and returns a
+// *RehydratedValidationRecordFields. To save database space, boulder no longer
+// stores the hostname and port fields in the validation record for an HTTP-01
+// challenge because that data can be rehydrated from the url field.
+func TransformURLIntoHostnameAndPort(input string) (*RehydratedValidationRecordFields, error) {
+	url, err := url.Parse(input)
+	if err != nil {
+		return &RehydratedValidationRecordFields{}, fmt.Errorf("could not parse url %v", input)
+	}
+
+	// A safety check for the scheme being either https or http is done in the
+	// VA prior to entering this transformation.
+	scheme := url.Scheme
+	hostname := url.Hostname()
+	if hostname == "" {
+		return &RehydratedValidationRecordFields{}, fmt.Errorf("could not parse hostname: %v", input)
+	}
+	port := url.Port()
+	if scheme == "https" && port == "" {
+		port = "443"
+	} else if scheme == "http" && port == "" {
+		port = "80"
+	}
+
```

Comment:
1. We can use `url.Scheme` directly here instead of allocating it to `scheme`.
2. We can `switch` on scheme and fit both of these into two tidy cases.
3. We can use a default case which will return an error if we're ever wrong about this assumption.
```suggestion
	port := url.Port()
	if port == "" {
		switch url.Scheme {
		case "https":
			port = "443"
		case "http":
			port = "80"
		default:
			// This should never happen since the VA should have already
			// checked the scheme.
			return "", "", fmt.Errorf("unknown scheme %q in URL %q", url.Scheme, vr.URL)
		}
```